### ISSUE TITLE
numerical stability problem

### ISSUE THREAD

**Issue by wuzirui** [nerfstudio-project/gsplat#248]
Hi! I've recently run into a numerical stability problem with the rasterization kernel as shown below:
I call the rasterization kernel with the following command:
![image](https://github.com/nerfstudio-project/gsplat/assets/7344146/a17c37a2-2c75-43cf-808a-7931b104a722)
I've checked the inputs and all the tensors are guaranteed to be not nan, but when I tried to calculate the gradient of `render_outputs` or `alpha` over gaussian params (e.g. `scales`), nan values appear:
![image](https://github.com/nerfstudio-project/gsplat/assets/7344146/fd78372a-c1bb-4102-a772-d752aa127932)

Could you please give me some hints on which part of the code may be numerically unstable so I can come up with some walkaround, or where I shall start to fix the problem? Thanks a lot!

**Comment by maturk** [comment]
Hi @wuzirui, one thing that came to mind, are you certain that all gaussians are visible by your current viewmat. If gaussians are not visible (which means they do not land withing the camera height/width after projection), they are not retained in mem and no gradients flow to them.

**Comment by wuzirui** [comment]
> 

Hi @maturk! Thanks for the reply, but if this is the case, the grad value for those gaussians would be zero instead of nans? For my data, the training can proceed ~several hundred steps and then crash (all gaussians turn into nans in 1~2 steps), so I guess it is not likely because of visibility issues?

**Comment by maturk** [comment]
hmm, what kind of dataset are you trying to train? @wuzirui, you can check the following `info.radii =< 0` and see if the radii that are 0 are the same as the ones with nan grad. if radii == 0, then it is not in view of the camera after projection.

**Comment by liruilong940607** [comment]
If the scales is super small, it could be numerically instable. Maybe try adding a small eps to the scales?
```
scales = torch.exp(scales_crop) + 1e-4
```

**Comment by wuzirui** [comment]
> hmm yes that is strange, what kind of dataset are you trying to train? I am wondering if there are any problems with projection. @wuzirui, you can check the following `info["radii"] =< 0` and see if the radii that are 0 are the same as the ones with nan grad.

Thanks! But as can be seen here, all the Gaussians have positive or zero radii, none of them are negative, so I guess this is not the case?
![image](https://github.com/nerfstudio-project/gsplat/assets/7344146/c74f7c96-9458-4cc7-850a-f4a45266fe1c)

**Comment by wuzirui** [comment]
> If the scales is super small, it could be numerically instable. Maybe try adding a small eps to the scales?
> 
> ```
> scales = torch.exp(scales_crop) + 1e-4
> ```

Hi @liruilong940607, thanks for the comment! The minimum scale input to the kernel is 0.0024 (shown below) so I guess this is also not the case? Also I pasted the range of all the input tensors FYI, thanks a lot!
![image](https://github.com/nerfstudio-project/gsplat/assets/7344146/25d6a5b9-1750-408d-a6f2-4e1d91aa7a88)

**Comment by quan5609** [comment]
Hi @wuzirui, did you figure out your problem?

**Comment by wuzirui** [comment]
> Hi @wuzirui, did you figure out your problem?

not yet, maybe later I need
 to look into the kernel

**Comment by npurson** [comment]
Similar issue here. I have switched from the official implementation to gsplat. The visualization of pre-trained gaussians yield same results for both module. However, the model is unconverged during training, and the grad norm is NaN.

**Comment by npurson** [comment]
> Similar issue here. I have switched from the official implementation to gsplat. The visualization of pre-trained gaussians yield same results for both module. However, the model is unconverged during training, and the grad norm is NaN.

BTW, the output of `torch.autograd.detect_analomy` is

```
'_FullyFusedProjectionPackedBackward' returned nan values in its 0th output.
```

**Comment by realimposter** [comment]
> _FullyFusedProjectionPackedBackward

Did you find a solution? I'm getting the same error

**Comment by npurson** [comment]
> > _FullyFusedProjectionPackedBackward
> 
> Did you find a solution? I'm getting the same error

Hi, I've looked the inputs and found that a small portion of the `scales` values were NaN due to precision errors when using fp16 for coordinate system transformations. This seems to be fine for the original diff-GS implementation, but causes problems with gsplat. Handling these NaN values resolved the issue for me.

**Comment by insomniaaac** [comment]
> > > _FullyFusedProjectionPackedBackward
> > 
> > 
> > Did you find a solution? I'm getting the same error
> 
> Hi, I've looked the inputs and found that a small portion of the `scales` values were NaN due to precision errors when using fp16 for coordinate system transformations. This seems to be fine for the original diff-GS implementation, but causes problems with gsplat. Handling these NaN values resolved the issue for me.

meet same issue, could you give some code snippets to solve this?

**Comment by RayYoh** [comment]
> Similar issue here. I have switched from the official implementation to gsplat. The visualization of pre-trained gaussians yield same results for both module. However, the model is unconverged during training, and the grad norm is NaN.

Hi, when I change to gsplat I find that sometimes it's easy to converge but sometimes not. 
Sometimes I also encounter Nan values. Have you fixed it?

**Comment by wuzirui** [comment]
Hey guys, I've solved the issue by simply scaling up the scene by a constant number:

all the location-related values (i.e. means, translation part of the camera pose...)

the weakest part is the scales, so a simple test is to see whether the nan issue persists if you clamp all the scales to a relatively larger number, say 0.01

**Comment by RayYoh** [comment]
> > > Similar issue here. I have switched from the official implementation to gsplat. The visualization of pre-trained gaussians yield same results for both module. However, the model is unconverged during training, and the grad norm is NaN.
> > 
> > 
> > Hi, when I change to gsplat I find that sometimes it's easy to converge but sometimes not. Sometimes I also encounter Nan values. Have you fixed it?
> 
> I found that it was due to training with fp16, where the original version can converge while gsplat can't.

Hi there. Actually, before feeding those parameters into gsplat I make them be fp32 like this:

![Image](https://github.com/user-attachments/assets/afa3a480-58b8-45f6-9efc-4f728567b7da)

Additionally, I render the depth map with render mode `RGB+D`, which may output inf value. I have not checked if they are inf or -inf...
